components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/device/DeviceEndpoint.java

```diff
@@ -0,0 +1,137 @@
+import org.apache.oltu.oauth2.common.exception.OAuthSystemException;
+import org.apache.oltu.oauth2.common.message.OAuthResponse;
+import org.wso2.carbon.identity.core.util.IdentityUtil;
+import org.wso2.carbon.identity.oauth.common.exception.InvalidOAuthClientException;
+import org.wso2.carbon.identity.oauth2.IdentityOAuth2Exception;
+import org.wso2.carbon.identity.oauth2.device.codegenerator.GenerateKeys;
+import org.wso2.carbon.identity.oauth2.device.constants.Constants;
+import org.wso2.carbon.identity.oauth2.device.dao.DeviceFlowPersistenceFactory;
+import org.wso2.carbon.identity.oauth2.device.errorcodes.DeviceErrorCodes;
+import org.wso2.carbon.identity.oauth2.device.model.DeviceFlowDO;
+
+import java.io.IOException;
+import java.util.UUID;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.Response;
+
+@Path("/device_authorize")
+public class DeviceEndpoint {
+
+    private static final Log log = LogFactory.getLog(DeviceEndpoint.class);
+
+    @POST
+    @Path("/")
+    @Consumes("application/x-www-form-urlencoded")
+    @Produces("application/json")
+    public Response authorize(@Context HttpServletRequest request, @Context HttpServletResponse response)
+            throws IOException, IdentityOAuth2Exception, InvalidOAuthClientException, OAuthSystemException {
+
+        String userCode = new GenerateKeys().getKey(6);
+        DeviceFlowDO deviceFlowDO = new DeviceFlowDO();
+        String deviceCode = UUID.randomUUID().toString();
+        String clientId = request.getParameter(Constants.CLIENT_ID);
+        String scope = request.getParameter(Constants.SCOPE);
+        String[] scopeSet = scope.split(" ");
+        deviceFlowDO.setScope(scopeSet);
+        String redirectionUri = IdentityUtil.getServerURL("/authenticationendpoint/device.do",
+                false, false);
+        String redirectionUriComplete = redirectionUri + "?user_code=" + userCode;
+        Long expiresIn = 3600000L;
+        int interval = 5000;
+        OAuthResponse errorResponse;
+
+        if (clientId != null) {
+            if (validateClientId(clientId)) {
+
+                DeviceFlowPersistenceFactory.getInstance().getDeviceFlowDAO().insertDeviceFlow(deviceCode, userCode,
+                        clientId, scope, expiresIn);
+
+                OAuthResponse deviceResponse =
+                        OAuthResponse.status(HttpServletResponse.SC_ACCEPTED).setParam(Constants.DEVICE_CODE,
+                                deviceCode).setParam(Constants.USER_CODE, userCode).setParam(Constants.VERIFICATION_URI,
+                                redirectionUri).setParam(Constants.VERIFICATION_URI_COMPLETE, redirectionUriComplete).
+                                setParam(Constants.EXPIRES_IN, String.valueOf(expiresIn / 1000))
```

Comment:
String.valueOf(expiresIn / 1000) Let's extract this into a method and use it for similar needs